tensorboard/db.py

```diff
@@ -20,160 +20,492 @@
+          formatted: events.out.tfevents.UNIX_TIMESTAMP.HOSTNAME[SUFFIX].
+      offset: The byte offset in the event log file *after* the last
+          successfully committed event record.
+    """
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE TABLE IF NOT EXISTS EventLogs (
+          rowid INTEGER PRIMARY KEY,
+          run_id INTEGER NOT NULL,
+          path VARCHAR(255) NOT NULL,
+          offset INTEGER NOT NULL
+        )
+      ''')
+
+  def create_event_logs_table_path_index(self):
+    """Uniquely indexes the (name, path) fields on the event_logs table."""
+    with self._cursor() as c:
+      c.execute('''\
+        CREATE UNIQUE INDEX IF NOT EXISTS EventLogsPathIndex
+        ON EventLogs (run_id, path)
+      ''')
+
+  def _cursor(self):
+    return contextlib.closing(self._db_conn.cursor())  # type: sqlite3.Cursor
+
+
+class Id(object):
+  """Utility class for unsigned fixed bit IDs."""
+
+  def __init__(self, name, bits):
+    """Creates new instance.
+
+    :type name: str
+    :type bits: int
+    """
+    if bits < 1:
+      raise ValueError('bits must be >0')
+    self.name = name
+    self.bits = bits
+    self.max = _mask(bits)
+
+  def check(self, x):
+    """Throws ValueError if x isn't in bit range.
+
+    :type x: int
+    """
+    _check_id(x, self.bits, self.name)
+
+  def generate(self):
+    """Generates a random ID in the bit range.
+
+    :rtype: int
+    """
+    return random.randint(0, self.max)
+
+
+class RowId(object):
+  """Utility class for bit-packed SQLite primary keys."""
+
+  def __init__(self, name, global_id, local_id):
+    """Creates new instance.
+
+    :type name: str
+    :type global_id: Id
+    :type local_id: Id
+    """
+    self.name = name
+    self.bits = global_id.bits + local_id.bits
+    if self.bits > 63:
+      raise ValueError('%s can not exceed 63 bits' % name)
+    self._global = global_id
+    self._local = local_id
+
+  def create(self, high, low):
+    """Creates a rowid from its global and local portions.
+
+    :type high: int
+    :type low: int
+    :rtype: int
+    """
+    self._global.check(high)
+    self._local.check(low)
+    return (high << self._local.bits) + low
+
+  def parse(self, rowid):
+    """Parses a rowid into its global and local portions.
+
+    :type rowid: int
+    :rtype: tuple[int, int]
+    """
+    _check_id(rowid, self.bits, self.name)
+    return rowid >> self._local.bits, rowid & self._local.max
+
+  def get_range(self, high):
+    """Returns an inclusive range of all possible y values.
+
+    This is used for the SQL BETWEEN operator.
+
+    :type high: int
+    :rtype: tuple[int, int]
+    """
+    return self.create(high, 0), self.create(high, self._local.max)
+
+
+def _check_id(id_, bits, name):
+  if id_ < 0:
+    raise ValueError('%s can not be a negative number: %d' % (name, id_))
+  if id_ > _mask(bits):
+    raise ValueError('%s must be a %d-bit number: %d' % (name, bits, id_))
+
+
+def _mask(bits):
```

Comment:
How about, <code>"""Return the highest integer that can be stored in &grave;bits&grave; unsigned bits."""</code>?